examples_tests/43.AllocatorTest/main.cpp

```diff
@@ -1,53 +1,219 @@
+		void* reservedSpace = _IRR_ALIGNED_MALLOC(reservedSize, _IRR_SIMD_ALIGNMENT);
+		auto poolAlctr = PoolAddressAllocator(reservedSpace, randOffset, randAlignOffset, randMaxAlign, randAddressSpaceSize, randBlockSz);
+
+		// randomly decide how many `multi_allocs` to do
+		const uint32_t multiAllocCnt = rng.getRandomNumber(1u, 5u); //range?
+		for (uint32_t i = 0u; i < multiAllocCnt; i++)
+		{
+			// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+			const uint32_t allocCntInMultiAlloc = rng.getRandomNumber(1u, 4096u); //?
+
+			for (size_t i = 0u; i < allocCntInMultiAlloc; ++i)
+			{
+				// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				perFrameData.sizes[i] = rng.getRandomNumber(1u, Traits::max_size(poolAlctr));
+				perFrameData.alignments[i] = rng.getRndMaxAlign();
+			}
+
+			Traits::multi_alloc_addr(poolAlctr, allocCntInMultiAlloc, perFrameData.outAddr, perFrameData.sizes, perFrameData.alignments);
+
+			// record all successful alloc addresses to the `core::vector`
+			if (perFrameData.outAddr[i] != PoolAddressAllocator::invalid_address)
+				results.push_back({ perFrameData.outAddr[i], perFrameData.sizes[i], perFrameData.alignments[i] });
+
+			randFreeAllocatedAddresses(poolAlctr);
+		}
+
+		// randomly choose between reset and freeing all `core::vector` elements
+		bool reset = static_cast<bool>(rng.getRandomNumber(0u, 1u));
+		if (reset)
+		{
+			poolAlctr.reset();
+		}
+		else
+		{
+			// free everything with a series of multi_free
+			while (results.size() != 0u)
+				randFreeAllocatedAddresses(poolAlctr);
+		}
+		 
+		_IRR_ALIGNED_FREE(reservedSpace);
+	}
+
+	// random dealloc function
+		// randomly decide how many calls to `multi_alloc` | ??
+			// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+				// if traits say address allocator supports arbitrary order free, then choose randomly, else choose most recently allocated
+	void randFreeAllocatedAddresses(PoolAddressAllocator& alctr)
+	{
+		//here I'm freeing previously generated addresses, is it ok?
+
+		if(results.size() == 0u)
+			return;
+
+		// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+		//TODO: random interval
+		const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size());
+
+		for (uint32_t i = 0u; i < addressesToFreeCnt; i++)
+		{
+			perFrameData.outAddr[i] = results[i].outAddr;
+			perFrameData.sizes[i] = results[i].size;
 		}
-		_IRR_ALIGNED_FREE(resSpc);
+
+		Traits::multi_free_addr(alctr, addressesToFreeCnt, perFrameData.outAddr, perFrameData.sizes);
+		results.erase(results.begin(), results.begin() + addressesToFreeCnt);
```

Comment:
The O(N^2) nature of this `erase` call will cripple us.

Reply:
You need:
- a dummy (NOOP) container for linear allocator
- a FILO container for stack allocator
- a container that supports random deletion and in-order insert in O(logn) or O(1)

Reply:
> The O(N^2) nature of this erase call will cripple us.

isn't it linear? https://en.cppreference.com/w/cpp/container/vector/erase

Reply:
for a single removal its linear O(N), now do N removals on a vector of N items will give you N^2 runtime